Rewrite FBX Importer to support Maya

RevoluPowered opened the pull request:
Adds the ability to load Autodesk FBX files, without requiring the SDK or thirdparty dependencies. We don't support 3DS max as of yet, due to a limitation of geometric transform. We will add support for this in the future, if required.

Previously due to FBX Pivot Transforms this was not possible, also inherit type (nodes which do not inherit parent scale, or do different multiplication orders) was previously complex to support and is required to support Maya fully.

We abstract most of Ref<PivotTransform> away and keep this simple for anyone who needs to make some edit to the importer. Pivot transform handles pivot's and also inherit type (scaling ignored, parent before scaling, etc)

Previously we had four layers, one for assimp one for godot, and one for the importer itself converting the data to godot, this removes assimp from the middle simplifiying the method to 

FBX Document -> Godot Importer (this is a direct importer now, and should be much more efficient in some cases over the previous importer)

Another advantage is we also only have a single set of core math types between FBX and Godot (Vector3, Vector2 and Transform), less conversions, less can go wrong.

This can be extended with as many features as we want now and there is now room to grow it into all the specific tools for Maya etc, or even possibly 3ds max.

**Why the decision to drop assimp?** 
- We didn't drop it entirely, we only use the FBX portion of the document parser.
- We moved all the internal assimp types like aiMatrix aiVector3D to Vector3D and Transform respectively.
- Multiple conversion layers caused issues which are harder to track down with three formats to handle it was harder to add the more complex FBX elements.

**FBX Importer with viable support for commercial FBX files**
- pivots (Pivot transforms fully supported, excluding 3DS Max as of now)
- inherit types (Segment Scale Compensation works ON or OFF; unity doesn't have this supported 🍰 )
- pre and post-rotation (orient bones and skeleton correctly same as original FBX / Modeller application)
- pivoted animations (Pivot transforms fully supported, excluding 3DS Max as of now)
- meshes
- normals
- vertex colors
- vertex weights
- node animations
- node locator animations
- bone animations
- blend shapes (as other importers use Normalize we use normalized blend shapes)
- file scaling


This work is sponsored by IMVU.

fire commented on modules/fbx_importer/SCsub:
```diff
@@ -0,0 +1,111 @@
+env_assimp.Append(CPPDEFINES=["OPENDDLPARSER_BUILD"])
+env_assimp.Append(CPPDEFINES=["ASSIMP_BUILD_NO_OWN_ZLIB"])
+env_assimp.Append(CPPDEFINES=["ASSIMP_BUILD_NO_EXPORT"])
+env_assimp.Append(CPPDEFINES=["ASSIMP_BUILD_NO_X_IMPORTER"])
```
Are these still necessary?

RevoluPowered replied:
yup

RevoluPowered replied:
I will strip out all of assimp except the FBX folder in a bit, maybe once we deliver next milestone.

fire commented on modules/fbx_importer/data/fbx_bone.h:
```diff
@@ -0,0 +1,96 @@
+struct FBXBone;
+
+struct FBXBone : Reference {
+	uint64_t parent_bone_id;
```
Uninitialized.

RevoluPowered replied:
I'm updating everything to be initialized by default wherever possible, there is a bug in no-except for Map in some cases with Vector3 being initialized too, might file this later on today.

fire commented on modules/fbx_importer/data/fbx_node.h:
```diff
@@ -43,91 +49,25 @@
-	uint32_t import_flags;
-};
+struct FBXNode : Reference, ModelAbstraction {
+	uint64_t current_node_id;
```
Uninitialized.

fire commented on modules/fbx_importer/editor_scene_importer_fbx.h:
```diff
@@ -0,0 +1,208 @@
+
+	struct ImportFormat {
+		Vector<String> extensions;
+		bool is_default;
```
Initialized.

RevoluPowered replied:
Fixed all cases I could see thanks!

RevoluPowered commented:
All cases found for the DOM have been initialized @fire let me know if you can see anymore.

I might actually split up the 'assimp' code into our single folder under modules since we are maintaining this ourselves in our codebase with our own types, so we don't have a third-party anymore

RevoluPowered commented:
More instances of uninitialized variables are in ASSIMP::FBX:: on DOM elements themselves. :D 

Will fix this when I get some time after blend shapes.

RevoluPowered commented:
I have rewritten render code to make it more reliable as it's now abstract enough to hanlde lines, points, quads and triangles :)

RevoluPowered commented on thirdparty/assimp/code/FBX/FBXMeshGeometry.h:
```diff
@@ -93,106 +96,72 @@ class MeshGeometry : public Geometry {
+		MapType map_type = MapType::none;
+		std::vector<T> data;
+		std::vector<int> index;
+	};
```
DOM should only read data not process it

RevoluPowered commented:
In the process of rewriting the MeshGeometry:: to make it only read the data rather than also manipulate it.

This patch disables all mesh rendering as we are gonna rewrite the importer in godot to handle the FBX polygon type properly, so we can also directly use indices.

This will resolve other issues too and make it testable.

NormalsW is not handled at the moment but this opens the opportunity to support it. :)

RevoluPowered commented:
I have fixed the bone, skin, node attribute mapping system in the importer, this now filters skin poses so they don't get populated for incorrect bones without any skinned mesh.

This fixes regressions found in models. It also allows for blender support again <3 and fixes vertex weight data.

I have also added file scaling to handle arbitrary file scaling.

I added Material mapping too, this is preliminary phase I have one fix to apply here to finish this entirely. 

I will leave the blend shapes and UVs fixes to @AndreaCatania for next week.

RevoluPowered commented:
https://www.youtube.com/watch?v=fgMmBeFmTSo&feature=youtu.be

fire commented on modules/fbx_importer/data/fbx_material.h:
```diff
@@ -115,39 +111,39 @@ struct FBXMaterial : Reference {
+		{ "3dsMax|Parameters|metalness_map", SpatialMaterial::TextureParam::TEXTURE_METALLIC },
+		{ "Maya|TEX_metallic_map|file", SpatialMaterial::TextureParam::TEXTURE_METALLIC },
+		{ "SpecularColor", SpatialMaterial::TextureParam::TEXTURE_METALLIC },
+		{ "Maya|SpecularTexture", SpatialMaterial::TextureParam::TEXTURE_METALLIC },
```
"SpecularColor" and "SpecularTexture" are not metallic textures. They're a combination of base color and roughness. Metallic is glossiness...

fire replied:
See https://github.com/godotengine/godot/pull/38438/files#diff-de1d3a22612435f4a0efc95d83e02952R1578

RevoluPowered replied:
these will be removed, the conversion here for each type of material type is not going to be added since there are 40-60 different properties to support

fire replied:
I believe SpecularColor, SpecularTexture is for the legacy texturing of FBX. As per our discussion, you mean that legacy materials will be removed? That makes sense, but you probably want to go into depth.

As per the discussion we mentioned there's legacy texturing and like 3+ sets of PBR texturing stacked.

RevoluPowered replied:
no, just incompatible mappings which require conversion, so some legacy stuff will remain; the point will be to keep compatibility for the basics (like color), specular is not something really that matters a lot in the grand scheme of things to getting this over the line for me, I just want to get this sorted and merged, then we can add whatever and then worry about all the finer detail conversion for the materials.

At best materials won't really produce the same results in different packages for me, so converting them seems pointless, my intention is to use these as placeholders for now, and actually spend time ensuring the mappings from maya are compatible, because IMHO they're not all going to be the same formula if they're called specular, they could have an FBX specific way of handling this in one system like GLTF2 but it seems like a guess to assume it's the same algorithm for specular, especially since FBX seems to be built on top of their own everything

fire commented:
I'll post the fox fbx that breaks this. 
[f993eb95eda14be3949d23309ff13d76.zip](https://github.com/godotengine/godot/files/4838129/f993eb95eda14be3949d23309ff13d76.zip)

https://sketchfab.com/3d-models/low-poly-fox-by-pixelmannen-animated-371dea88d7e04a76af5763f2a36866bc

RevoluPowered commented:
> I'll post the fox fbx that breaks this.
> [f993eb95eda14be3949d23309ff13d76.zip](https://github.com/godotengine/godot/files/4838129/f993eb95eda14be3949d23309ff13d76.zip)
> 
> https://sketchfab.com/3d-models/low-poly-fox-by-pixelmannen-animated-371dea88d7e04a76af5763f2a36866bc
> 
> ![image](https://user-images.githubusercontent.com/32321/85880458-e9912a80-b790-11ea-9fbe-2602139c774a.png)

@fire can we make a tracker for files that don't load properly?

there is gonna be a lot of files we need to make fixes for during stablization

fire commented:
Started here. https://github.com/godotengine/godot/issues/39853

fire commented on modules/fbx_importer/data/fbx_mesh_data.cpp:
```diff
@@ -539,8 +513,8 @@ HashMap<int, T> FBXMeshData::extract_per_vertex_data(
 					ERR_FAIL_COND_V_MSG(vertex_index < 0, (HashMap<int, T>()), "FBX file corrupted: #ERR8");
-					ERR_FAIL_COND_V_MSG(vertex_index >= p_vertex_count, (HashMap<int, T>()), "FBX file seems corrupted: #ERR9.")
-					ERR_FAIL_COND_V_MSG(p_fbx_data.index[polygon_vertex_index] < 0, (HashMap<int, T>()), "FBX file seems corrupted: #ERR10.")
+					ERR_FAIL_COND_V_MSG(vertex_index >= p_vertex_count, (HashMap<int, T>()), "FBX fileseems  corrupted: #ERR9.")
```
Typo in message.

fileseems -> file seems

AndreaCatania replied:
Err 9 is just a way for us to know the kind of error, in case someone reports the issue.
In case of the file is corrupted, or seems so, there is no good way to tell what's going on to the user, nor is possible for him fix that. So the best message is "The file is corrupted".

So it's much better just say that, and return a number so in case of issue report, we know where the corruption is.

fire commented on modules/fbx_importer/data/fbx_mesh_data.cpp:
```diff
@@ -85,69 +110,83 @@ struct SurfaceData {
+			uvs_1_raw);
+
+	// Make sure that from this moment on the mesh_geometry is no used anymore.
+	// This is a safety step, becuase the mesh_geometry data are no more valid
```
becuase -> because

fire commented on modules/fbx_importer/data/fbx_mesh_data.cpp:
```diff
@@ -320,14 +360,174 @@ MeshInstance *FBXMeshData::create_fbx_mesh(const ImportState &state, const Assim
+				} else if (uvs->size() > 0) {
+					// No uv, this is strange, just take the first and duplicate.
+					(*this_vert_poly_uv) = *uvs->getptr(*uvs->next(nullptr));
+					WARN_PRINT("No UVs for this polygon, while there is no default and some other polygons has it. This FBX file may be corrupted.");
```
"No UVs for this polygon, while there is no default and some other polygons has it. This FBX file may be corrupted."

This sentence is awkward.

RevoluPowered replied:
Fixed but I'm still not happy with the fix I did.

fire commented on modules/fbx_importer/tools/import_utils.cpp:
```diff
@@ -127,21 +127,23 @@ Transform get_unscaled_transform(const Transform &p_initial, real_t p_scale) {
-	double x = 0.0;
-	double y = 0.0;
-	double z = 0.0;
+	// Using long double to make sure that normal is computed foreven really tiny objects.
```
foreven -> for even

fire commented on modules/fbx_importer/data/fbx_mesh_data.cpp:
```diff
@@ -547,79 +794,86 @@ const int FBXMeshData::count_polygons(const std::vector<int> &p_polygon_indices)
-					ERR_FAIL_COND_V_MSG(p_fbx_data.index[polygon_vertex_index] >= (int)p_fbx_data.data.size(), (HashMap<int, T>()), "FBX file seems corrupted: #ERR11.")
-					aggregate_vertex_data[vertex_index].push_back(p_fbx_data.data[p_fbx_data.index[polygon_vertex_index]]);
+					ERR_FAIL_COND_V_MSG(vertex_index < 0, (HashMap<int, R>()), "FBX file corrupted: #ERR8");
+					ERR_FAIL_COND_V_MSG(vertex_index >= p_vertex_count, (HashMap<int, R>()), "FBX fileseems  corrupted: #ERR9.")
```
"FBX fileseems  corrupted: #ERR9."

->

 "FBX file seems corrupted: #ERR9."

RevoluPowered commented:
@fire @AndreaCatania co-authored and rebased, we need to review code after we add blend shape anim tracks.

Not sure how well blender is working now, we should test.

RevoluPowered commented:
Bugs left:
- inherit type rrs is not working
- dds loading
- i need to fix blend shape anim tracks then we can upstream after code review

fire commented:
Can we rename the fbx module to modules/fbx, so I have less changes to do. It'll make my life implementing exporting fbx easier because merging will keep diffs instead of a folder name change. You can ignore this suggestion if you want.

RevoluPowered commented:
> Can we rename the fbx module to modules/fbx, so I have less changes to do. It'll make my life implementing exporting fbx easier because merging will keep diffs instead of a folder name change. You can ignore this suggestion if you want.

Ideally we only do exporting for 4.0, I will make a new PR for 4.0 with this change though and we can do the fbx module itself there :) if that suits?

RevoluPowered commented:
Spoke to @fire we will refactor to FBX/ tomorrow

fire commented:
You probably meant "fbx/" like the other modules, but wanted to check.

RevoluPowered commented:
Yeah fbx/

On Wed, 22 Jul 2020 at 21:06, K. S. Ernest (iFire) Lee <
notifications@github.com> wrote:

> You probably meant "fbx/" like the other modules, but wanted to check.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/39418#issuecomment-662668827>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFWZYQOFUGSYOILJSRZDBDR45BEJANCNFSM4NZRDIEA>
> .
>

RevoluPowered commented:
I had to edit the structure for the third time to allow the static-checks to be valid.

RevoluPowered commented:
Need to fix the core edit I made and revert it.

Geequlim commented:
Why not support 3dx-max?
Almost all modelers and animators are using 3dmax in the game industry in my city.

RevoluPowered commented:
> Why not support 3dx-max?
> Almost all modelers and animators are using 3dmax in the game industry in my city.

To implement 3DS Max we need to integrate geometric pivots which require overly intense computation on import. I have had them working in a previous iteration but we removed them when we had to rewrite for the 5th time.

geometric_pivot in PivotTransform -> is basically only applied to the current node, so all the children nodes have to be affine_inversed based on the parent geometric transform, this means the performance seriously suffers in a horrible way, per node you're iterating per node, per geometric pivot, affining per child and then, applying the final global transform, it's very prohibitively expensive. 

I'm not opposed to adding it we can do that, but we're gonna merge the first iteration first and implement fixes along the way, so we can do incremental fixes for support.

RevoluPowered commented:
The best approach to applying geometric_translation/rotation/scaling is to apply it at the end if nonzero is found for geometric pivots

RevoluPowered commented:
Removed new flags which didn't make much sense, I discussed them with @reduz and he explained it's possible to do already.

qarmin commented:
Importing empty fbx file exported from blender cause leak
Project(just open to import) - [AAA.zip](https://github.com/godotengine/godot/files/5016611/AAA.zip)

Leak:
```
Direct leak of 3760 byte(s) in 94 object(s) allocated from:
    #0 0x7f0e27123f17 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.6+0xb1f17)
    #1 0x2504ef7 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:364
    #2 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
    #3 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
    #4 0x25060c2 in Assimp::FBX::TokenizeBinary(std::vector<Assimp::FBX::Token const*, std::allocator<Assimp::FBX::Token const*> >&, char const*, unsigned long) thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:442
    #5 0x1e2563c in EditorSceneImporterFBX::import_scene(String const&, unsigned int, int, List<String, DefaultAllocator>*, Error*) modules/fbx/editor_scene_importer_fbx.cpp:133
    #6 0x3b5fb6e in ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_scene.cpp:1318
    #7 0x3255da4 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1798
    #8 0x325c780 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1996
    #9 0x3239d0e in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:589
    #10 0x3244efe in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1174
    #11 0x32635b7 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pobrane/godot-feature-fbx_final_stage/bin/godot.x11.tools.64s+0x32635b7)
    #12 0x6d96f2f in Object::notification(int, bool) core/object.cpp:932
    #13 0x4a8cf86 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:985
    #14 0x4a87d92 in SceneTree::idle(float) scene/main/scene_tree.cpp:525
    #15 0x143af52 in Main::iteration() main/main.cpp:2107
    #16 0x13bd359 in OS_X11::run() platform/x11/os_x11.cpp:3233
    #17 0x138f8a5 in main platform/x11/godot_x11.cpp:56
    #18 0x7f0e265e30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

```

RevoluPowered commented:
> Importing empty fbx file exported from blender cause leak
> Project(just open to import) - [AAA.zip](https://github.com/godotengine/godot/files/5016611/AAA.zip)
> 
> Leak:
> 
> ```
> Direct leak of 3760 byte(s) in 94 object(s) allocated from:
>     #0 0x7f0e27123f17 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.6+0xb1f17)
>     #1 0x2504ef7 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:364
>     #2 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
>     #3 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
>     #4 0x25060c2 in Assimp::FBX::TokenizeBinary(std::vector<Assimp::FBX::Token const*, std::allocator<Assimp::FBX::Token const*> >&, char const*, unsigned long) thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:442
>     #5 0x1e2563c in EditorSceneImporterFBX::import_scene(String const&, unsigned int, int, List<String, DefaultAllocator>*, Error*) modules/fbx/editor_scene_importer_fbx.cpp:133
>     #6 0x3b5fb6e in ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_scene.cpp:1318
>     #7 0x3255da4 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1798
>     #8 0x325c780 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1996
>     #9 0x3239d0e in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:589
>     #10 0x3244efe in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1174
>     #11 0x32635b7 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pobrane/godot-feature-fbx_final_stage/bin/godot.x11.tools.64s+0x32635b7)
>     #12 0x6d96f2f in Object::notification(int, bool) core/object.cpp:932
>     #13 0x4a8cf86 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:985
>     #14 0x4a87d92 in SceneTree::idle(float) scene/main/scene_tree.cpp:525
>     #15 0x143af52 in Main::iteration() main/main.cpp:2107
>     #16 0x13bd359 in OS_X11::run() platform/x11/os_x11.cpp:3233
>     #17 0x138f8a5 in main platform/x11/godot_x11.cpp:56
>     #18 0x7f0e265e30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
> ```

Thanks, just to confirm what options do you use to detect the leak?

qarmin commented:
`scons p=x11 -j6 use_lsan=yes` should be enough on Linux but I'm using usually `scons p=x11 -j6 use_asan=yes` which also finds invalid memory usage(leak sanitizer is enabled by default on address sanitizer)
On macOS and windows probably needs to be enabled settings on Visual Studio or XCode

RevoluPowered commented:
> `scons p=x11 -j6 use_lsan=yes` should be enough on Linux but I'm using usually `scons p=x11 -j6 use_asan=yes` which also finds invalid memory usage(leak sanitizer is enabled by default on address sanitizer)
> On macOS and windows probably needs to be enabled settings on Visual Studio or XCode

Thanks! yeah just wanted to set up the same testing environment thank you so much for reporting this issue :)

RevoluPowered commented:
> > `scons p=x11 -j6 use_lsan=yes` should be enough on Linux but I'm using usually `scons p=x11 -j6 use_asan=yes` which also finds invalid memory usage(leak sanitizer is enabled by default on address sanitizer)
> > On macOS and windows probably needs to be enabled settings on Visual Studio or XCode
> 
> Thanks! yeah just wanted to set up the same testing environment thank you so much for reporting this issue :)

I believe the problem is the Assimp::FBX::Document is not deleted. I will make a patch now and see if this is resolved

RevoluPowered commented:
> Importing empty fbx file exported from blender cause leak
> Project(just open to import) - [AAA.zip](https://github.com/godotengine/godot/files/5016611/AAA.zip)
> 
> Leak:
> 
> ```
> Direct leak of 3760 byte(s) in 94 object(s) allocated from:
>     #0 0x7f0e27123f17 in operator new(unsigned long) (/lib/x86_64-linux-gnu/libasan.so.6+0xb1f17)
>     #1 0x2504ef7 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:364
>     #2 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
>     #3 0x2505585 in ReadScope thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:389
>     #4 0x25060c2 in Assimp::FBX::TokenizeBinary(std::vector<Assimp::FBX::Token const*, std::allocator<Assimp::FBX::Token const*> >&, char const*, unsigned long) thirdparty/assimp_fbx/FBXBinaryTokenizer.cpp:442
>     #5 0x1e2563c in EditorSceneImporterFBX::import_scene(String const&, unsigned int, int, List<String, DefaultAllocator>*, Error*) modules/fbx/editor_scene_importer_fbx.cpp:133
>     #6 0x3b5fb6e in ResourceImporterScene::import(String const&, String const&, Map<StringName, Variant, Comparator<StringName>, DefaultAllocator> const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) editor/import/resource_importer_scene.cpp:1318
>     #7 0x3255da4 in EditorFileSystem::_reimport_file(String const&) editor/editor_file_system.cpp:1798
>     #8 0x325c780 in EditorFileSystem::reimport_files(Vector<String> const&) editor/editor_file_system.cpp:1996
>     #9 0x3239d0e in EditorFileSystem::_update_scan_actions() editor/editor_file_system.cpp:589
>     #10 0x3244efe in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1174
>     #11 0x32635b7 in EditorFileSystem::_notificationv(int, bool) (/home/rafal/Pobrane/godot-feature-fbx_final_stage/bin/godot.x11.tools.64s+0x32635b7)
>     #12 0x6d96f2f in Object::notification(int, bool) core/object.cpp:932
>     #13 0x4a8cf86 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:985
>     #14 0x4a87d92 in SceneTree::idle(float) scene/main/scene_tree.cpp:525
>     #15 0x143af52 in Main::iteration() main/main.cpp:2107
>     #16 0x13bd359 in OS_X11::run() platform/x11/os_x11.cpp:3233
>     #17 0x138f8a5 in main platform/x11/godot_x11.cpp:56
>     #18 0x7f0e265e30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
> ```

> `scons p=x11 -j6 use_lsan=yes` should be enough on Linux but I'm using usually `scons p=x11 -j6 use_asan=yes` which also finds invalid memory usage(leak sanitizer is enabled by default on address sanitizer)
> On macOS and windows probably needs to be enabled settings on Visual Studio or XCode

I have resolved the leak on my side, could you please confirm it's gone?

qarmin commented:
I'm still having leak
You need each time remove `.import` folder to check leak, because it happens only when importing model.

RevoluPowered commented:
> I'm still having leak
> You need each time remove `.import` folder to check leak, because it happens only when importing model.

I can't reproduce any-more on my end, I tried removing import and also removing the import file and reimporting nothing, tried several times, did you pull?

![image](https://user-images.githubusercontent.com/748770/89205433-a4da8980-d5af-11ea-817d-1b9715cf66dc.png)

qarmin commented:
Sorry, indeed in latest commit there isn't any leak.
I'm not sure how I get this leak, but now everything works fine.

You can add `DRI_PRIME=0` before command to prevent showing false leaks, like I see in your screenshot like:
`DRI_PRIME=0 godot.x11.tools.64s -e`

RevoluPowered commented:
> Sorry, indeed in latest commit there isn't any leak.
> I'm not sure how I get this leak, but now everything works fine.
> 
> You can add `DRI_PRIME=0` before command to prevent showing false leaks, like I see in your screenshot like:
> `DRI_PRIME=0 godot.x11.tools.64s -e`

No need to apologize it happens and I have done this before myself too, I am just happy it's fixed, I didn't know that I needed DRI_PRIME what is that?

qarmin commented:
This command switches management on/off(in case =0 - off) of hybrid GPU - https://wiki.archlinux.org/index.php/PRIME
Here is a Godot bug report with leaking memory with PRIME detection - https://github.com/godotengine/godot/issues/29844

fire commented:
@RevoluPowered Here's the html log. 
[2020-08-13-godot-fbx.zip](https://github.com/godotengine/godot/files/5071115/2020-08-13-godot-fbx.zip)

```
pvs-studio-analyzer credentials fire@example.com key
git clone https://github.com/RevoluPowered/godot.git -b feature/fbx_final_stage
cd godot
pvs-studio-analyzer analyze -o godot-static-check.log -j6
plog-converter -t fullhtml -o 2020-08-13-godot-fbx godot-static-check.log
# Zip and send
```

RevoluPowered committed:
FBX Importer Rewrite

Co-authored-by: Gordon MacPherson <gordon@gordonite.tech>
Co-authored-by: Andrea Catania <info@andreacatania.com>
Co-authored-by: K. S. Ernest (iFire) Lee <ernest.lee@chibifire.com>

This is a complete rewrite of the importer. It will give more deterministic behaviour and has been sponsored by IMVU inc, over 1 year has gone into the development of this importer to remove the burden of the FBX SDK.

This was my project for 1 entire year and I really enjoyed the opportunity to add to Godot.

Along the road of implementing fixes we implemented fbx pivots, animations and inheritance type handling, which in most cases works properly.

We have implemented animation and mesh skinning too this should work out of the box, if there are issues let us know.

It's designed so that you can expand this with ease, and fix bugs easily too.

It can import from Autodesk Maya and import into Godot, with pivots.

There are bits we could polish but for now this is good enough.

This was sponsored by IMVU, and a special thanks to everyone who supported this project.

Signed-off-by: Gordon MacPherson <gordon@gordonite.tech>

Additional fixes made before upstreaming:
- fixed memory leak
- ensure consistent ordering on mac linux and windows for fbx tree. (very important for material import to be deterministic)
- disabled incorrect warnings for fbx_material
- added compatibility code for /RootNode/ so compat is not broken
- Optimise FBX - directly import triangles
- remove debug messages
- add messages for mesh id, mesh re-import is sometimes slow and we need to know what mesh is being worked on
- correct culling settings on materials

Xrayez commented:
There are link errors if you try to compile with `scons module_fbx_enabled=no`, see #39196 (I'm currently compiling with `module_assimp_enabled=no` in 3.2).

```
[ 98%] Linking Program        ==> bin\godot.windows.tools.64.exe
[ 98%]    Creating library bin\godot.windows.tools.64.lib and object bin\godot.windows.tools.64.exp
tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::deg2rad(struct Vector3 const &)" (?deg2rad@ImportUtils@@SA?AUVector3@@AEBU2@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::rad2deg(struct Vector3 const &)" (?rad2deg@ImportUtils@@SA?AUVector3@@AEBU2@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static class Basis __cdecl ImportUtils::EulerToBasis(enum Assimp::FBX::Model::RotOrder,struct Vector3 const &)" (?EulerToBasis@ImportUtils@@SA?AVBasis@@W4RotOrder@Model@FBX@Assimp@@AEBUVector3@@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::BasisToEuler(enum Assimp::FBX::Model::RotOrder,class Basis const &)" (?BasisToEuler@ImportUtils@@SA?AUVector3@@W4RotOrder@Model@FBX@Assimp@@AEBVBasis@@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
bin\godot.windows.tools.64.exe : fatal error LNK1120: 4 unresolved externals
scons: *** [bin\godot.windows.tools.64.exe] Error 1120
scons: building terminated because of errors.
```

You'll need to conditionally compile tests by defining `FBX_ENABLED` in SCons I think, see #35963 where module defines are implemented in `master`. Or just remove the tests...

RevoluPowered commented:
> There are link errors if you try to compile with `scons module_fbx_enabled=no`, see #39196 (I'm currently compiling with `module_assimp_enabled=no` in 3.2).
> 
> ```
> [ 98%] Linking Program        ==> bin\godot.windows.tools.64.exe
> [ 98%]    Creating library bin\godot.windows.tools.64.lib and object bin\godot.windows.tools.64.exp
> tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::deg2rad(struct Vector3 const &)" (?deg2rad@ImportUtils@@SA?AUVector3@@AEBU2@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
> tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::rad2deg(struct Vector3 const &)" (?rad2deg@ImportUtils@@SA?AUVector3@@AEBU2@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
> tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static class Basis __cdecl ImportUtils::EulerToBasis(enum Assimp::FBX::Model::RotOrder,struct Vector3 const &)" (?EulerToBasis@ImportUtils@@SA?AVBasis@@W4RotOrder@Model@FBX@Assimp@@AEBUVector3@@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
> tests.windows.tools.64.lib(test_fbx.windows.tools.64.obj) : error LNK2019: unresolved external symbol "public: static struct Vector3 __cdecl ImportUtils::BasisToEuler(enum Assimp::FBX::Model::RotOrder,class Basis const &)" (?BasisToEuler@ImportUtils@@SA?AUVector3@@W4RotOrder@Model@FBX@Assimp@@AEBVBasis@@@Z) referenced in function "bool __cdecl TestFBX::test_rotation(struct Vector3,enum Assimp::FBX::Model::RotOrder)" (?test_rotation@TestFBX@@YA_NUVector3@@W4RotOrder@Model@FBX@Assimp@@@Z)
> bin\godot.windows.tools.64.exe : fatal error LNK1120: 4 unresolved externals
> scons: *** [bin\godot.windows.tools.64.exe] Error 1120
> scons: building terminated because of errors.
> ```
> 
> You'll need to conditionally compile tests by defining `FBX_ENABLED` in SCons I think, see #35963 where module defines are implemented in `master`. Or just remove the tests...

Fixed :) We removed the old tests for now, no point in modifying them right now, more info is here:
- https://github.com/godotengine/godot/issues/40659#issuecomment-677781744

They're backed up and saved to be ported into 4.0 rather than added to 3.2 since doctest is not backported.

RevoluPowered committed:
fix importer failing to import 7700 files, and correct args for version

also fixed some cleanup notes

RevoluPowered committed:
working uv2?

RevoluPowered committed:
Fix UV2 coordinates and various problems with FBX
- Document no longer uses unordered maps
- Removed some usages of &GetRequiredToken replaced with safe *GetRequiredToken() function
- Added parser debugging
- Added ERR_FAIL_CONDS for unsupported mesh formats (we can add these later super easy to do now)
- Add memory debugging for the Tokens and the TokenParser to make it safe
- Add memory initialisation to mesh.cpp surface_tool.h and mesh.h
- Initialize boolean flags properly
- Refactored to correct naming for the fbx_mesh_data.h so you know what data you are working on
- Disabled corruption caused by the FIXME:
- Fixed document reading indexes and index_to_direct vs indexes mode

akien-mga commented:
Superseded by #41979.
